crypto/fipsmodule/service_indicator/service_indicator.c

```diff
@@ -174,6 +174,115 @@ void HMAC_verify_service_indicator(const EVP_MD *evp_md) {
   }
 }
 
+void DigestSign_verify_service_indicator(const EVP_MD_CTX *ctx) {
+  if(ctx->pctx->pmeth->pkey_id == EVP_PKEY_RSA ||
+      ctx->pctx->pmeth->pkey_id == EVP_PKEY_RSA_PSS) {
+    // SHA1 and 1024 bit keys are not approved for RSA signature generation.
+    // SHA2-224, SHA2-256, SHA2-384, SHA2-512 with 2048, 3072 and 4096 bit keys
+    // are approved for signature generation.
+    if (EVP_PKEY_size(ctx->pctx->pkey) == 256 ||
+        EVP_PKEY_size(ctx->pctx->pkey) == 384 ||
+        EVP_PKEY_size(ctx->pctx->pkey) == 512) {
+      switch (ctx->digest->type) {
+        case NID_sha224:
+        case NID_sha256:
+        case NID_sha384:
+        case NID_sha512:
+          FIPS_service_indicator_update_state();
+          break;
+        default:
+          break;
+      }
+    }
+  } else if(ctx->pctx->pmeth->pkey_id == EVP_PKEY_EC){
+    // Not the best way to write this, but the delocate parser for ARM/clang can't
+    // recognize || if statements for this. Moving the curve check before the
+    // digest check also causes delocate parser failures.
+    // TODO: Update the delocate parser to be able to recognize a more readable
+    // version of this.
+    //
+    // SHA1 is not approved for ECDSA signature generation.
+    // Curves P-224, P-256, P-384 and P-521 with SHA2-224, SHA2-256, SHA2-384 and
+    // SHA2-512 are approved for signature generation.
+    int curve_name = EC_GROUP_get_curve_name(ctx->pctx->pkey->pkey.ec->group);
+    switch (ctx->digest->type) {
```

Comment:
```
static int ECDSA_Digest_curve_approval(const int curve_name) {
  // Curves P-224, P-256, P-384 and P-521 are approved for signature generation.
  switch(curve_name) {
    case NID_secp224r1:
    case NID_X9_62_prime256v1:
    case NID_secp384r1:
    case NID_secp521r1:
      return 1;
    default:
      return 0;
  }
}


static int DigestSign_hash_approval(const int nid) {
  // SHA2-224, SHA2-256, SHA2-384 and SHA2-512 approved for signature
  // generation.
  switch(nid) {
    case NID_sha224:
    case NID_sha256:
    case NID_sha384:
    case NID_sha512:
      return 1;
    default:
      return 0;
  }
}
```

I tried changing this function by changing the multiple switch checks into static functions for each (like the above ones), but it just throws me into the another loop of trying out multiple combinations to get the build past the delocater parser. 
I think we should tackle all the weird check methods all at once once the delocater is updated, otherwise it's only becoming a game of guessing what gets past a broken parser in the current state.

Reply:
Ok np - SIM?

Reply:
I'll update the existing one with more documentation and the new cases.